<?xml version="1.0"?>

<robot name="ebamk2" xmlns:xacro="http://www.ros.org/wiki/xacro">
 <xacro:include filename="$(find eezybot)/urdf/gazebo.xacro" />
 <xacro:ros_control prefix="/"/>
  <xacro:macro name="cuboid_inertia_def" params="x y z mass">
    <inertia 
      iyy="${mass*(z*z+x*x)/12.0}"
      ixx="${mass*(y*y+z*z)/12.0}"
      izz="${mass*(x*x+y*y)/12.0}"
      ixy="0.0" iyz="0.0" ixz="0.0"/>
  </xacro:macro>

<link name="world"/>
<link name="base_link"/>

<!-- BASE LINK -->

  <joint name="fixed" type="fixed">
    <parent link="world"/>
    <child link="base_link"/>
    <origin xyz="0 0 0" rpy="0 0 0" />
  </joint>

  <link
    name="link_0">
    <visual>
      <origin xyz="0 0 0" rpy="0 0 0" />
      <geometry>
        <mesh filename="package://eezybot/meshes/base_link.STL" />
      </geometry>
      <material name="">
        <color rgba="1 0 0 1" />
      </material>
    </visual>
    <collision>
      <origin xyz="0 0 0.0275" rpy="0 0 0" />
      <geometry>
        <box size="0.055 0.055 0.055"/>
      </geometry>
    </collision>
    <inertial>
      <origin xyz="0 0 0.0275" rpy="0 0 0"/>
      <mass value="5"/>
      <xacro:cuboid_inertia_def x="0.055" y="0.055" z="0.055" mass="5"/>
    </inertial>
  </link>

  <joint name="base_link_to_link_0" type="fixed">
    <parent link="base_link"/>
    <child link="link_0"/>
    <origin xyz="0.0 0.0 0.0" rpy="0 0 0" />
  </joint>
<!-- JOINT 1 HIP-->
  <joint name="joint_1" type="revolute">
    <origin xyz="0 0 0.05" rpy="0 0 0" />
    <parent link="link_0" />
    <child link="link_1" />
    <axis xyz="0 0 1" />
    <limit effort="10" velocity="0.50"
      lower="-0.7854" upper="0.7854"/>
  </joint>

<!-- LINK 1 HIP-->
  <link
    name="link_1">
    <visual>
      <origin
        xyz="0 -0.04 0" rpy="0 0 0" />
      <geometry>
        <mesh filename="package://eezybot/meshes/link_1.STL" />
      </geometry>
      <material name="">
        <color rgba="0 0.275 0 1" />
      </material>
    </visual>
    <collision>
      <origin
        xyz="0 0 0.04" rpy="0 0 0" />
      <geometry>
        <box size="0.032 0.052 0.0415"/>
      </geometry>
    </collision>
    <inertial>
      <origin xyz="0 0 0.04" rpy="0 0 0"/>
      <mass value="0.15"/>
      <xacro:cuboid_inertia_def x="0.032" y="0.052" z="0.0415" mass="0.15"/>
    </inertial>
  </link>

<!-- JOINT 2 SHOULDER -->
  <joint name="joint_2" type="revolute">
    <origin xyz="0 -0.005 0.037" rpy="1.578 0 0" />
    <parent link="link_1" />
    <child link="link_2" />
    <axis xyz="0 0 1" />
    <limit effort="10" velocity="0.40"
      lower="-0.80" upper="0.35"/>
  </joint>

<!-- LINK 2 SHOULDER -->
  <link name="link_2">
    <visual>
      <origin xyz="0 0 0.0075" rpy="3.14159 0 0" />
      <geometry>
        <mesh filename="package://eezybot/meshes/link_2.STL" />
      </geometry>
      <material name="">
        <color rgba="0 0 1 1" />
      </material>
    </visual>
    <collision>
      <origin xyz="${(-0.021-0)/2.0} ${(0.135+0.005)/2.0} ${0.02+(-0.0105-0.037)/2.0}" rpy="3.14159 0 ${0.14888994}" />
      <geometry>
        <box size="${0.01} ${0.135+0.005} ${0.01}"/>
      </geometry>
    </collision>
    <inertial>
      <origin xyz="${(-0.021-0)/2.0} ${(0.135+0.005)/2.0} ${0.02+(-0.0105-0.037)/2.0}" rpy="3.14159 0 ${0.14888994}" />
      <mass value="0.12"/>
      <xacro:cuboid_inertia_def x="0.01" y="0.135" z="0.01" mass="0.12"/>
    </inertial>
  </link>

  <!-- VIRTUAL MIMIC JOINT 2-->
  <!--joint name="joint2v" type="revolute">
    <origin xyz="-0.021 0.135 -0.0105" rpy="1.570796325 0 0" />
    <parent link="link_2"/>
    <child link="link_2v"/>
    <axis xyz="0 1 0"/>
    <limit effort="30" velocity="1.0" lower="-3.14159" upper="3.14159" />
    <mimic joint="joint_2" multiplier="-1" offset="0" />
  </joint-->

 
  <joint name="joint_3" type="revolute">
    <origin xyz="-0.021 0.135 -0.0105" rpy="1.570796325 0 0" />
    <parent link="link_2"/>
    <child link="link_3"/>
    <axis xyz="0 1 0"/>
    <limit effort="30" velocity="0.5" lower="${-0.49}" upper="1.22" />
    <!--mimic joint="joint_2" multiplier="-1" offset="0" /-->
  </joint>

<!-- JOINT 3 ELBOW -->
  <!--joint name="joint_3" type="revolute">
    <origin xyz="0 0 0" rpy="0 0 0" />
    <parent link="link_2v" />
    <child link="link_3" />
    <axis xyz="0 1 0" />
    <limit effort="0" velocity="0"
      lower="-0.77" upper="0.45"/>
  </joint-->

<!-- LINK 3 ELBOW -->
   <link name="link_3">
    <visual>
      <origin xyz="0 0.0335  0.002" rpy="1.570796325 0 0" />
      <geometry>
        <mesh filename="package://eezybot/meshes/link_3.STL" />
      </geometry>
      <material name="">
        <color rgba="1 0 0 1" />
      </material>
    </visual>
    <collision>
      <origin xyz="${(0.1459)/2.0} ${(0.0203)/2.0} ${0.0185/2.}" rpy="1.570796325 ${-0.126126098371} 0" />
      <geometry>
        <box size="${0.1459} 0.01 0.01" />
      </geometry>
    </collision>
    <inertial>
      <origin xyz="${(0.1459)/2.0} ${(0.0203)/2.0} ${0.0185/2.}" rpy="1.570796325 ${-0.126126098371} 0" />
      <mass value="0.07"/>
      <xacro:cuboid_inertia_def x="0.1459" y="0.01" z="0.01" mass="0.07"/>
    </inertial>
  </link>

  <!-- JOINT 4 CLAW SUPPORT -->
  <joint name="joint_4" type="revolute">
    <origin xyz="0.1459 0.0185 0.0203" rpy="0 1.570796325 3.14159" />
    <parent link="link_3" />
    <child link="link_4" />
    <axis xyz="0 1 0" />
    <limit effort="10" velocity="0.4" lower="-0.5" upper="1"/>
    <!--<mimic joint="joint_3" multiplier="1" offset="0" />-->
  </joint>

  <!-- LINK 4 CLAW SUPPORT -->
  <link name="link_4">
    <visual>
      <origin xyz="-0.006 0 -0.0394" rpy="0 0 0" />
      <geometry>
        <mesh filename="package://eezybot/meshes/link_4.STL" scale="0.001 0.001 0.001" />
      </geometry>
      <material name="">
        <color rgba="0 0.275 0 1" />
      </material>  
    </visual>
    <collision>
      <origin xyz="0.015 0.015 -0.02" rpy="0 0 0" />
      <geometry>
        <box size="0.03 0.03 0.04" />
      </geometry>
    </collision>
    <inertial>
      <origin xyz="0.015 0.015 -0.02" rpy="0 0 0" />
      <mass value="0.03"/>
      <xacro:cuboid_inertia_def x="0.03" y="0.03" z="0.04" mass="0.03"/>
    </inertial>
  </link>

  <!-- JOINT 5 CLAW BASE -->
  <joint name="joint_5" type="fixed">
    <origin xyz="-0.0095 0.001 -0.033" rpy="-0 1.570796325 0" />
    <parent link="link_4" />
    <child link="claw_base" />
    <axis xyz="1 0 0" />
  </joint>

  <!-- LINK 4 CLAW BASE -->
  <link name="claw_base">
    <visual>
      <origin xyz="0 0 0" rpy="0 0 0" />
      <geometry>
        <mesh filename="package://eezybot/meshes/link_5.STL" scale="0.001 0.001 0.001" />
      </geometry>
      <material name="">
        <color rgba="0 0 1 1" />
      </material>
    </visual>
    <collision>
      <origin xyz="0 0 0" rpy="0 0 0" />
      <geometry>
        <mesh filename="package://eezybot/meshes/link_5.STL" scale="0.001 0.001 0.001" />
      </geometry>
    </collision>
    <inertial>
      <origin xyz="0 0 0" rpy="0 0 0" />
      <mass value="0.03"/>
      <xacro:cuboid_inertia_def x="0.03" y="0.02" z="0.005" mass="0.03"/>
    </inertial>
  </link>

  <!-- FINGERS -->
  <joint name="joint_6" type="revolute">
    <origin xyz="0.0196 0.021 -0.004" rpy="0 0 0" />
    <parent link="claw_base"/>
    <child link="claw_l"/>
    <axis xyz="0 0 1" />
    <limit effort="0" lower="0" upper="0.65" velocity="1.0" />
  </joint>

  <joint name="joint_7" type="revolute">
    <origin xyz="0.0196 0.0040 -0.004" rpy="0 0 0" />
    <parent link="claw_base"/>
    <child link="claw_r"/>
    <axis xyz="0 0 -1" />
    <limit effort="0" lower="0" upper="0.65" velocity="1.0" />
    <mimic joint="joint_6" multiplier="1.0" offset="0" />
  </joint>

  <link name="claw_l">
    <visual>
      <origin xyz="0.0505 0.000 0" rpy="0 0 2.9" />
      <geometry>
        <mesh filename="package://eezybot/meshes/claw_l.STL" scale="0.001 0.001 0.001" />
      </geometry>
    </visual>
    <collision>
      <origin xyz="0.0505 0.000 0" rpy="0 0 2.9" />
      <geometry>
        <mesh filename="package://eezybot/meshes/claw_l.STL" scale="0.001 0.001 0.001" />
      </geometry>
    </collision>
    <inertial>
      <origin xyz="0.00 0.000 0" rpy="0 0 2.9" />
      <mass value="0.01"/>
      <xacro:cuboid_inertia_def x="0.047" y="0.005" z="0.005" mass="0.01"/>
    </inertial>
  </link>

  <link name="claw_r">
    <visual>
      <origin xyz="0.0488 0.0102 0" rpy="0 0 3.14159" />
      <geometry>
        <mesh filename="package://eezybot/meshes/claw_r.STL" scale="0.001 0.001 0.001" />
      </geometry>
    </visual>
    <collision>
      <origin xyz="0.0488 0.0102 0" rpy="0 0 3.14159" />
      <geometry>
        <mesh filename="package://eezybot/meshes/claw_r.STL" scale="0.001 0.001 0.001" />
      </geometry>
    </collision>
    <inertial>
      <origin xyz="0.0235 0.0 0" rpy="0 0 3.14159" />
      <mass value="0.01"/>
      <xacro:cuboid_inertia_def x="0.047" y="0.005" z="0.005" mass="0.01"/>
    </inertial>
  </link>
<!-- TRANSMISSIONS -->
  <transmission name="trans_1">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="joint_1">
      <hardwareInterface>EffortJointInterface</hardwareInterface>
    </joint>
    <actuator name="servo1">
      <hardwareInterface>EffortJointInterface</hardwareInterface>
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>
  <transmission name="trans_2">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="joint_2">
      <hardwareInterface>EffortJointInterface</hardwareInterface>
    </joint>
    <actuator name="servo2">
      <hardwareInterface>EffortJointInterface</hardwareInterface>
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>
  <transmission name="trans_3">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="joint_3">
      <hardwareInterface>EffortJointInterface</hardwareInterface>
    </joint>
    <actuator name="servo3">
      <hardwareInterface>EffortJointInterface</hardwareInterface>
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>
  <transmission name="trans_4">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="joint_4">
      <hardwareInterface>EffortJointInterface</hardwareInterface>
    </joint>
    <actuator name="servo4">
      <hardwareInterface>EffortJointInterface</hardwareInterface>
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>
</robot>
